README
The following instructions can be found on the Alternate Login GitHub Wiki: https://github.com/mjimeyg/alternate-login/wiki

*** INSTALLATION INSTRUCTIONS ****

Download AutoMOD from [http://www.phpbb.com/mods/automod/ phpBB AutoMOD]

Upload the contents of the 'upload' folder into the root of your forum.

Download UMIL from [http://www.phpbb.com/mods/umil/ phpBB UMIL]

Upload the contents of the root folder to the root of your forum.

Visit the install file in your browser i.e. http://www.example.com/myforum/install/install.php

Follow the on screen instructions.

Delete the install folder or rename it.

Log into your forum's ACP and go to the AutoMOD page.

Upload the Alternate Login zip file using the form provided.

Click the install link for the Alternate Login MOD on the AutoMOD page.


Follow the on screen instructions.

MAKE SURE TO CLICK THE LINK IN THE SUMMARY OF THE INSTALLATION!!!


At this point Alternate Login is installed and you need to move on to Post-Installation Instructions.


**** Post installation instructions ****

For three of the providers (Facebook, Twitter and Windows Live) you will need to visit the developers site for each provider and create an 'App' for your own site. These 'Apps' are FREE!

OpenID can be initialised straight off the bat with no need for additional settings or creating 'Apps'. You can skip to the last section to see details on or to initialise it.

== Setup Facebook ==

Go to [https://developers.facebook.com/|Facebook Developer's site].

Create an App, fill in as much of the details as you can. If you have never done this before you might be asked to verify you are who you say you are via SMS text messaging. It's relatively painless.

Once you have created your App, take note of the App ID and App Secret.

Go to your forum ACP and navigate to the ConSof page, Facebook Manager section.

You now need to fill in various fields, the following are mandatory fields:

:: a. Facebook App ID - You get this from the Facebook App you just created.

:: b. Facebook Secret - You get this from the Facebook App you just created.

:: c.Site Domain - This is the domain of your forum. If your forum is located at http://www.example.com/phpBB/index.php then your domain is example.com.

The Page URL is optional and is only needed if you enable the Facebook Like Box. For more information about Facebook Pages see: [http://www.facebook.com/about/pages Facebook Pages]

Finally, select the various plugins you want displayed on your index page and click submit.

Now continue with the other providers or move onto the final steps.

== Setup Twitter ==

Go to the [https://dev.twitter.com/apps Twitter Developers Site].

Sign in and create an App by following the on screen instructions. Fill in as much information as you can.

:: '''NOTE!''' The callback should point to the al_tw_connect.php file. (i.e. http://www.example.com/phpBB/alternatelogin/al_tw_connect.php)

3.Note down the Consumer Key, Consumer Secret and Callback URL.

Go to your forum ACP and navigate to the ConSof page, Twitter Manager section.

Now fill in ALL the fields using the values you noted down in step 3.

Finally click submit.

Now continue with the other providers or move onto the #Final Steps.

== Setup Windows Live ==

Go to the [http://msdn.microsoft.com/en-us/live/default.aspx Windows Live Developer's Site].

Create an Application following the on screen instructions.

:: '''NOTE!''' The callback should point to the al_wl_connect.php file. (i.e. http://www.example.com/phpBB/alternatelogin/al_wl_connect.php)

Note down the Client ID and Client Secret and Callback URL.

Go to your forum ACP and navigate to the ConSof page, Windows Live Manager section.

Now fill in ALL the fields using the values you noted down in step 3.

Finally click submit.

Now move onto the #Final Steps.

== Final Steps ==  

Click on the Alternate Login Manager section of the ConSof ACP page.

Enable each of the providers you wish to use and have set up for. (Remember that OpenID does not require additional settings.)

Click Submit.

You should now be able to log in with these providers.

changelog.txt
2012-06-25		Michael Goonawardena    <webmaster@consof.com>

	* alternatelogin/al_fb_login.php : Removed erroneous call to add non-existant language file.



2012-06-07		Michael Goonawardena    <webmaster@consof.com>

	* includes/ucp/ucp_profile.php : Added function_exists call to check before including functions_alternatelogin.php file.


2012-05-23		Michael Goonawardena    <webmaster@consof.com>

	* al_fb_connect.php : Removed lines 467 - 481 and corresponding if brackets.  COPPA and email registration not required for quick accounts.
	
	
2012-05-13		Michael Goonawardena    <webmaster@consof.com>

	* al_fb_connect.php  : Reversed the lines at 57 and 58 to strip relative path from $return_to_page.
	* styles/prosilver/template/js/openid-en.js : Added entry for Steam.
	* styles/subsilver2/template/js/openid-en.js : Added entry for Steam.
	* styles/prosilver/template/js/openid-jquery.js : Changed value of 'no_sprite' to true.
	* styles/subsilver2/template/js/openid-jquery.js : Changed value of 'no_sprite' to true.


2012-04-26		Michael Goonawardena    <webmaster@consof.com>

	* styles/prosilver/template/index_body.html : Added missing AL_FB_LIKE_BOX variable to first IF statement.
	* styles/subsilver2/template/index_body.html : Added missing AL_FB_LIKE_BOX variable to first IF statement.
	* styles/prosilver/template/login_body.html : Moved Alternate Login buttons outside of login form to fix conflict with OI.
	* styles/subsilver2/template/login_body.html : Moved Alternate Login buttons outside of login form to fix conflict with OI.
	* styles/prosilver/template/login_body.html : Added Redirect variable to login urls.
	* styles/subsilver2/template/login_body.html : Added Redirect variable to login urls.
	* functions.php : Added template variable to assign redirect to login urls.
	* functions_alternatelogin.php : Edited get_fb_access_token() function to pass redirect url.
	* adm/index.php : Removed alterations to the file. File now remains unchanged on install.
	* al_fb_connect.php : Added meta_refresh function call to redirect after successful login and pass redirect page through authentication.
	* al_oi_connect.php : Added meta_refresh function call to redirect after successful login.
	* al_tw_connect.php : Added meta_refresh function call to redirect after successful login.
	* al_wl_connect.php : Added meta_refresh function call to redirect after successful login.
	* includes/acp/acp_alternatelogin.php : Added function call to Facebook settings to increase max_name_chars if Quick Accounts is enabled.


2012-03-13		Michael Goonawardena    <webmaster@consof.com>

	* al_fb_connect.php : Completely overhauled.
	* al_oi_connect.php : Completely overhauled.
	* al_tw_connect.php : Completely overhauled.
	* al_wl_connect.php : Completely overhauled.
	* facebook_registration : Added password fields.
	* ucp_register.php : Removed alteration of passwords.
	* ucp_register.html : Removed changes to password fields.


2012-02-14		Michael Goonawardena    <webmaster@consof.com>

	* styles/prosilver/template/images.large/steam.gif : Added Steam image.
	* styles/prosilver/template/js/openid-uk.js : Added Steam Provider.
	* functions_alternatelogin.php : corrected all add_log function calls.
	* functions_alternatelogin.php : Added 'global $user' declaration to get_fb_data() function.


2012-02-12		Michael Goonawardena    <webmaster@consof.com>

	* functions_alternatelogin.php : Replaced AL_MYSPACE_LOGIN and AL_MYSPACE_PROFILE constants with AL_OPENID_LOGIN and AL_OPENID_PROFILE respectively.
	* includes/acp/acp_alternatelogin.php : Added code to retrieve and set al_oi_login database entry.
	* al_oi_connect.php : Added file to handle OpenID accounts.
	* alternatelogin/openid/openid.php : Added 3rd party OpenID library.
	* acp_alternatelogin.html : Added radio buttons for OpenID.
	* styles/prosilver/template/index_body.php : Added OpenID login code.
	* styles/subsilver2/template/index_body.php : Added OpenID login code.
	* styles/prosilver/template/login_body.php : Added OpenID login code.
	* styles/subsilver2/template/login_body.php : Added OpenID login code.
	* info_acp_alternatelogin.php : Added entries for OPENID.
	* info_ucp_alternatelogin.php : Added entries for OPENID.

root/alternatelogin/al_fb_connect.php
<?php
/*
	COPYRIGHT 2009 Michael J Goonawardena

	This file is part of ConSof Alternate Login.

    ConSof Alternate Login is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    ConSof Alternate Login is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with ConSof Alternate Login.  If not, see <http://www.gnu.org/licenses/>.*/


// Basic setup of phpBB variables.
define('IN_PHPBB', true);
$phpbb_root_path = (defined('PHPBB_ROOT_PATH')) ? PHPBB_ROOT_PATH : '../';
$phpEx = substr(strrchr(__FILE__, '.'), 1);

// Load include files.
include($phpbb_root_path . 'common.' . $phpEx);
include_once($phpbb_root_path . 'includes/functions_user.' . $phpEx);
include_once($phpbb_root_path . 'includes/functions_alternatelogin.' . $phpEx);	// Custom Alternate Login functions.

// Set up a new user session.
$user->session_begin();
$auth->acl($user->data);
$user->setup('ucp');
$user->add_lang('mods/info_acp_alternatelogin');	// Global Alternate Login language file.
$user->add_lang('mods/info_ucp_alternatelogin');


// Make sure that Facebook login is enabled for this site.
if($config['al_fb_login'] == 0)
{
	// Inform the user that this feature is unavailable
	trigger_error(sprintf($user->lang['AL_LOGIN_UNAVAILABLE'], $user->lang['FACEBOOK']));
}

$return_to_page = request_var('return_to_page', base64_encode("{$phpbb_root_path}index.{$phpEx}"));

$access_token = get_fb_access_token($return_to_page);

if(!$access_token)
{
    add_log('critical', $user->data['user_id'], 'FB_ERROR_ACCESS_TOKEN');
    trigger_error($user->lang['FB_ERROR_ACCESS_TOKEN']);
}
$return_to_page = base64_decode($return_to_page);

$return_to_page = str_replace("../", "", $return_to_page);
$return_to_page = str_replace("./", "", $return_to_page);

// Store the access token for use with this session.
        $sql_array = array(
            'session_fb_access_token'   => $access_token,
        );

        $sql = "UPDATE " . SESSIONS_TABLE . " SET " . $db->sql_build_array('UPDATE', $sql_array) . " WHERE session_id='" . $user->data['session_id'] . "'";
        
        $db->sql_query($sql);
        
//echo 'token:' . print_r($token_url);
$graph_url = "https://graph.facebook.com/me?" . $access_token;


$fb_user = json_decode(get_fb_data($graph_url));

//echo("Hello " . $fb_user->name);
//print_r($fb_user);
// Check to see if we have a valid Facebook user.
if(!$fb_user)
{
    add_log('critical', $user->data['user_id'], 'FB_ERROR_USER');
    // Inform the user that we couldn't get their Facebook Id.
    trigger_error(sprintf($user->lang['FB_ERROR_USER'], $user->lang['FACEBOOK']));
}
$user->lang_name = substr($fb_user->locale, 0, 2);
// Select the user_id from the Alternate Login user data table which has the same Facebook Id.



$sql = 'SELECT user_id, username, user_password, user_passchg, user_pass_convert, user_email, user_type, user_login_attempts
		FROM ' . USERS_TABLE . "
		WHERE al_fb_id = " . $fb_user->id;
        

// Execute the query.
$result = $db->sql_query($sql);

// Retrieve the row data.
$row = $db->sql_fetchrow($result);

// Free up the result handle from the query.
$db->sql_freeresult($result);

// Check to see if we found a user_id with the associated Facebook Id.
if ($row)   // User is registered already, let's log him in!
{
        
	$old_session_id = $user->session_id;

        if ($admin)
        {
                global $SID, $_SID;

                $cookie_expire = time() - 31536000;
                $user->set_cookie('u', '', $cookie_expire);
                $user->set_cookie('sid', '', $cookie_expire);
                unset($cookie_expire);

                $SID = '?sid=';
                $user->session_id = $_SID = '';
        }

        $result = $user->session_create($row['user_id'], $admin, $autologin, $viewonline);
        
        // Successful session creation
        if ($result === true)
        {
                // If admin re-authentication we remove the old session entry because a new one has been created...
                if ($admin)
                {
                        // the login array is used because the user ids do not differ for re-authentication
                        $sql = 'DELETE FROM ' . SESSIONS_TABLE . "
                                WHERE session_id = '" . $db->sql_escape($old_session_id) . "'
                                AND session_user_id = {$row['user_id']}";
                        $db->sql_query($sql);
                }

                // Store the access token for use with this session.
                $sql_array = array(
                    'session_fb_access_token'   => $access_token,
                );

                $sql = "UPDATE " . SESSIONS_TABLE . " SET " . $db->sql_build_array('UPDATE', $sql_array) . " WHERE session_id='" . $user->data['session_id'] . "'";

                $db->sql_query($sql);
                $data = array();
                // Update the stored data such as profile and signatures.  Avatar is a dynamic field and doesn't require changing.

                if($user->data['al_fb_profile_sync'])
                {

                    $graph_url = "https://graph.facebook.com/me?" . $access_token;

                    $fb_user = json_decode(get_fb_data($graph_url));

                    $data['user_website']                    = (!$fb_user->website) ? '' : $fb_user->website;
                    $data['user_from']                   = (!$fb_user->location->name) ? '' : $fb_user->location->name;
                    $data['user_occ']                 = (!$fb_user->work[0]->employer->name) ? '' : $fb_user->work[0]->employer->name;
                    $bday = explode('/', $fb_user->birthday);
                    $data['user_birthday']              = sprintf('%2d-%2d-%4d', $bday[1], $bday[0], $bday[2]);

                }

                if($user->data['al_fb_status_sync'])
                {
                    include($phpbb_root_path . 'includes/message_parser.' . $phpEx);
                    $graph_url = "https://graph.facebook.com/me/statuses?" . $access_token;

                    $fb_user = json_decode(get_fb_data($graph_url));

                    $signature = $fb_user->data[0]->message;

                    $enable_bbcode                      = ($config['allow_sig_bbcode']) ? (bool) $user->optionget('sig_bbcode') : false;
                    $enable_smilies                     = ($config['allow_sig_smilies']) ? (bool) $user->optionget('sig_smilies') : false;
                    $enable_urls                        = ($config['allow_sig_links']) ? (bool) $user->optionget('sig_links') : false;

                    $message_parser = new parse_message($signature);

                    // Allowing Quote BBCode
                    $message_parser->parse($enable_bbcode, $enable_urls, $enable_smilies, $config['allow_sig_img'], $config['allow_sig_flash'], true, $config['allow_sig_links'], true, 'sig');

                    $data['user_sig']                   = (string) $message_parser->message;
                    $data['user_options']               = $user->data['user_options'];
                    $data['user_sig_bbcode_uid']	= (string) $message_parser->bbcode_uid;
                    $data['user_sig_bbcode_bitfield']	= $message_parser->bbcode_bitfield;
                }

                if($user->data['al_fb_profile_sync'] || $user->data['al_fb_status_sync'])
                {
                    $sql = 'UPDATE ' . USERS_TABLE . '
                            SET ' . $db->sql_build_array('UPDATE', $data) . '
                            WHERE user_id = ' . $user->data['user_id'];

                    $db->sql_query($sql);
                }
                
                meta_refresh(3, "$phpbb_root_path$return_to_page");
                trigger_error(sprintf($user->lang['LOGIN_SUCCESS'] . "<br /><br />" . sprintf($user->lang['RETURN_PAGE'], "<a href='{$phpbb_root_path}{$return_to_page}'>", "</a>")));
                
                redirect("{$phpbb_root_path}index.$phpEx");
        }
        else
        {
            trigger_error($user->lang['LOGIN_FAILED']);
        }
        

        

}

$sql = 'SELECT user_id, username, user_password, user_passchg, user_pass_convert, user_email, user_type, user_login_attempts
		FROM ' . USERS_TABLE . "
		WHERE user_email = '" . mysql_escape_string($fb_user->email) . "'";
        

// Execute the query.
$result = $db->sql_query($sql);

// Retrieve the row data.
$row = $db->sql_fetchrow($result);

// Free up the result handle from the query.
$db->sql_freeresult($result);

if($row)
{
    $old_session_id = $user->session_id;

        if ($admin)
        {
                global $SID, $_SID;

                $cookie_expire = time() - 31536000;
                $user->set_cookie('u', '', $cookie_expire);
                $user->set_cookie('sid', '', $cookie_expire);
                unset($cookie_expire);

                $SID = '?sid=';
                $user->session_id = $_SID = '';
        }

        $result = $user->session_create($row['user_id'], $admin, $autologin, $viewonline);

        // Store the access token for use with this session.
        $sql_array = array(
            'session_fb_access_token'   => $access_token,
        );

        $sql = "UPDATE " . SESSIONS_TABLE . " SET " . $db->sql_build_array('UPDATE', $sql_array) . " WHERE session_id='" . $user->data['session_id'] . "'";

        $db->sql_query($sql);
        
        // Successful session creation
        if ($result === true)
        {
                // If admin re-authentication we remove the old session entry because a new one has been created...
                if ($admin)
                {
                        // the login array is used because the user ids do not differ for re-authentication
                        $sql = 'DELETE FROM ' . SESSIONS_TABLE . "
                                WHERE session_id = '" . $db->sql_escape($old_session_id) . "'
                                AND session_user_id = {$row['user_id']}";
                        $db->sql_query($sql);
                }

                $sql_array = array(
                    'al_fb_id'      => $fb_user->id,
                    'al_wl_id'      => 0,
                    'al_tw_id'      => 0,
                    'al_oi_id'      => 0,
                );

                // Prepare the query to update the users Alternate Login record.
                $sql = 'UPDATE ' . USERS_TABLE
                . " SET " . $db->sql_build_array('UPDATE', $sql_array)
                . " WHERE user_id='{$user->data['user_id']}'";


                // Execute the query.
		$result = $db->sql_query($sql);

                if(!$result)
		{
			trigger_error($user->lang['AL_PHPBB_DB_FAILURE']);
		}
                else
		{
			trigger_error(sprintf($user->lang['LOGIN_SUCCESS'] . "<br /><br />" . sprintf($user->lang['RETURN_INDEX'], "<a href='{$phpbb_root_path}index.php'>", "</a>")));
		}
        }
        else
        {
            trigger_error($user->lang['LOGIN_FAILED']);
        }
}
else
{
	// No user was registered with the associate Facebook Id.
	// We need to see if they are anonymous.
	// If they are then that means they might want to register.
	// We will check to see if they wish to register.
	if($user->data['user_id'] == ANONYMOUS)
	{
            if(!$config['al_fb_quick_accounts'])
            {
                // Check to make sure the email is not already registered.
                $sql_array = array(
                    'SELECT'    => 'COUNT(user_email) AS user_email',
                    'FROM'      => array(USERS_TABLE => 'u'),
                    'WHERE'     => "user_email ='" . mysql_escape_string($fb_user->email) . "'",
                );
                
                $sql = $db->sql_build_query('SELECT', $sql_array);
                
                $result = $db->sql_query($sql);
                
                $email_registered = (int)$db->sql_fetchfield('user_email');
                
                if($email_registered > 0)
                {
                    trigger_error($user->lang['EMAIL_TAKEN_EMAIL']);
                }
		
                // Most of this code comes straight out of ucp_register.php
                $message = 'TERMS_OF_USE_CONTENT';
                $title = 'TERMS_USE';
                $user->lang_name = substr($fb_user->locale, 0, 2);
                if (empty($user->lang[$message]))
                {
                        if ($user->data['is_registered'])
                        {
                                redirect(append_sid("{$phpbb_root_path}index.$phpEx"));
                        }

                        login_box();
                }

                $template->set_filenames(array(
                        'body'		=> 'ucp_agreement.html')
                );

                // Disable online list
                page_header($user->lang[$title], false);

                $s_hidden_fields = array(	'al_login' 		=> 1,
                                                                        'al_login_type'	=> AL_FACEBOOK_LOGIN,
                                                                        'al_fb_user'	=> $fb_user->id
                );

                $add_lang                       = '&int=' . $fb_user->locale;
                $coppa				= (isset($_REQUEST['coppa'])) ? ((!empty($_REQUEST['coppa'])) ? 1 : 0) : false;
                        if ($coppa === false && $config['coppa_enable'])
                        {
                                $now = getdate();
                                $coppa_birthday = $user->format_date(mktime($now['hours'] + $user->data['user_dst'], $now['minutes'], $now['seconds'], $now['mon'], $now['mday'] - 1, $now['year'] - 13), $user->lang['DATE_FORMAT']);
                                unset($now);

                                $template->assign_vars(array(
                                        'L_COPPA_NO'		=> sprintf($user->lang['UCP_COPPA_BEFORE'], $coppa_birthday),
                                        'L_COPPA_YES'		=> sprintf($user->lang['UCP_COPPA_ON_AFTER'], $coppa_birthday),

                                        'U_COPPA_NO'		=> append_sid("{$phpbb_root_path}ucp.$phpEx", 'mode=register&amp;coppa=0' . $add_lang),
                                        'U_COPPA_YES'		=> append_sid("{$phpbb_root_path}ucp.$phpEx", 'mode=register&amp;coppa=1' . $add_lang),

                                        'S_SHOW_COPPA'		=> true,
                                        'S_HIDDEN_FIELDS'	=> build_hidden_fields($s_hidden_fields),
                                        'S_UCP_ACTION'		=> append_sid("{$phpbb_root_path}ucp.$phpEx", 'mode=register' . $add_lang),
                                ));
                        }
                        else
                        {
                                $template->assign_vars(array(
                                        'L_TERMS_OF_USE'	=> sprintf($user->lang['TERMS_OF_USE_CONTENT'], $config['sitename'], generate_board_url()),

                                        'S_SHOW_COPPA'		=> false,
                                        'S_REGISTRATION'	=> true,
                                        'S_HIDDEN_FIELDS'	=> build_hidden_fields($s_hidden_fields),
                                        'S_UCP_ACTION'		=> append_sid("{$phpbb_root_path}alternatelogin/al_fb_registration.$phpEx", 'mode=register' . $add_lang . $add_coppa),
                                        )
                                );
                        }

                page_footer();
            }
            else 
            {
                $data = array(
                    'username'		=> $fb_user->name,
                    'email'			=> strtolower($fb_user->email),
                    'email_confirm'		=> strtolower($fb_user->email),
                    'lang'                   => substr($fb_user->locale, 0, 2),
                    'tz'			=> (float) $fb_user->timezone,
                );
                
                $validate_username = validate_username($data['username']);

                if($validate_username)
                {
                    trigger_error($user->lang[$validate_username . '_USERNAME'] . ' <br /><br /><a href="' . $phpbb_root_path . '/alternatelogin/al_fb_registration.' . $phpEx . '?mode=register">' . $user->lang['BACK_TO_PREV'] . "</a>");            
                }

                $new_password = $fb_reg_data['user_id'] . $config['al_fb_key'] . $config['al_fb_secret'];


                $data['new_password'] = $new_password;
                $data['password_confirm'] = $new_password;
                add_log('critical', $user->data['user_id'], 'FB Password', $new_password);
                $error = validate_data($data, array(
                    'username'			=> array(
                                                        array('string', false, $config['min_name_chars'], $config['max_name_chars']),
                                                        array('username', '')),

                    'email'                     => array(
                                                        array('string', false, 6, 60),
                                                        array('email')),
                    'email_confirm'		=> array('string', false, 6, 60),
                    'tz'			=> array('num', false, -14, 14),
                    'lang'			=> array('match', false, '#^[a-z_\-]{2,}$#i'),
                ));


        
                // DNSBL check
                if ($config['check_dnsbl'])
                {
                    if (($dnsbl = $user->check_dnsbl('register')) !== false)
                    {
                        $error[] = sprintf($user->lang['IP_BLACKLISTED'], $user->ip, $dnsbl[1]);
                    }
                }

                if (!sizeof($error))
                {
                    if ($data['new_password'] != $data['password_confirm'])
                    {
                        $error[] = $user->lang['NEW_PASSWORD_ERROR'];
                    }

                    if ($data['email'] != $data['email_confirm'])
                    {
                        $error[] = $user->lang['NEW_EMAIL_ERROR'];
                    }
                }

                if (!sizeof($error))
                {
                    $server_url = generate_board_url();

                    // Which group by default?
                    $group_name = ($coppa) ? 'REGISTERED_COPPA' : 'REGISTERED';

                    $sql = 'SELECT group_id
                            FROM ' . GROUPS_TABLE . "
                            WHERE group_name = '" . $db->sql_escape($group_name) . "'
                                    AND group_type = " . GROUP_SPECIAL;
                    $result = $db->sql_query($sql);
                    $row = $db->sql_fetchrow($result);
                    $db->sql_freeresult($result);

                    if (!$row)
                    {
                        trigger_error('NO_GROUP');
                    }

                    $group_id = $row['group_id'];

                    if (($coppa ||
                            $config['require_activation'] == USER_ACTIVATION_SELF ||
                            $config['require_activation'] == USER_ACTIVATION_ADMIN) && $config['email_enable'])
                    {
                        $user_actkey = gen_rand_string(10);
                        $key_len = 54 - (strlen($server_url));
                        $key_len = ($key_len < 6) ? 6 : $key_len;
                        $user_actkey = substr($user_actkey, 0, $key_len);

                        $user_type = USER_INACTIVE;
                        $user_inactive_reason = INACTIVE_REGISTER;
                        $user_inactive_time = time();
                    }
                    else
                    {
                        $user_type = USER_NORMAL;
                        $user_actkey = '';
                        $user_inactive_reason = 0;
                        $user_inactive_time = 0;
                    }
                    $bday = explode('/', $fb_user->birthday);
                    $user_row = array(
                        'username'				=> $data['username'],
                        'user_password'			=> phpbb_hash($data['new_password']),
                        'user_email'			=> $data['email'],
                        'group_id'				=> (int) $group_id,
                        'user_timezone'			=> (float) $data['tz'],
                        'user_dst'				=> $is_dst,
                        'user_lang'				=> $data['lang'],
                        'user_type'				=> $user_type,
                        'user_actkey'			=> $user_actkey,
                        'user_ip'				=> $user->ip,
                        'user_regdate'			=> time(),
                        'user_inactive_reason'	=> $user_inactive_reason,
                        'user_inactive_time'	=> $user_inactive_time,
                        'al_fb_id'              => $fb_user->id,
                        'user_avatar_type'      => AVATAR_REMOTE,
                        'user_avatar_width'     => 100,
                        'user_avatar_height'    => 100,
                        'user_avatar'           => 'https://graph.facebook.com/' . $fb_user->id . '/picture?type=normal',
                        'al_fb_avatar_sync'     => 1,
                        'al_fb_profile_sync'    => 1,
                    
                        'user_website'                    => (!$fb_user->website) ? '' : $fb_user->website,
                        'user_from'                   => (!$fb_user->location->name) ? '' : $fb_user->location->name,
                        'user_occ'                 => (!$fb_user->work[0]->employer->name) ? '' : $fb_user->work[0]->employe->name,

                        'user_birthday'              => sprintf('%2d-%2d-%4d', $bday[1], $bday[0], $bday[2]),
                    );
                    
                    if ($config['new_member_post_limit'])
                    {
                        $user_row['user_new'] = 1;
                    }
                    
                    // Register user...
                    $user_id = user_add($user_row, $cp_data);

                    // This should not happen, because the required variables are listed above...
                    if ($user_id === false)
                    {
                        trigger_error('NO_USER', E_USER_ERROR);
                    }
                    
                    redirect("{$phpbb_root_path}/alternatelogin/al_fb_connect.{$phpEx}");
                }
                else
                {
                    trigger_error(implode('<br />', $error));
                }
            }
        }
}

?>
